src/cargo/core/compiler/standard_lib.rs

```diff
@@ -75,13 +75,8 @@ pub fn resolve_std<'gctx>(
     let src_path = detect_sysroot_src_path(target_data)?;
     let std_ws_manifest_path = src_path.join("Cargo.toml");
     let gctx = ws.gctx();
-    // TODO: Consider doing something to enforce --locked? Or to prevent the
-    // lock file from being written, such as setting ephemeral.
     let mut std_ws = Workspace::new(&std_ws_manifest_path, gctx)?;
-    // Don't require optional dependencies in this workspace, aka std's own
-    // `[dev-dependencies]`. No need for us to generate a `Resolve` which has
-    // those included because we'll never use them anyway.
-    std_ws.set_require_optional_deps(false);
```

Comment:
Why do we remove this? Did it cause any issue?

Reply:
> In order for Cargo to actually attempt to write the std lockfile we need to resolve with optional_deps again. Optional packages aren't actually built.

I guess it is for this?

Reply:
https://github.com/rust-lang/cargo/blob/15fbd2f607d4defc87053b8b76bf5038f2483cf4/src/cargo/ops/resolve.rs?plain=1#L289-L294

Okay it is this one. 

I am unsure if removing this is the correct approach. `require_optional_deps` affects how we resolve dependencies. And with that std starts respecting [path override](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#paths-overrides).

https://github.com/rust-lang/cargo/blob/15fbd2f607d4defc87053b8b76bf5038f2483cf4/src/cargo/ops/resolve.rs?plain=1#L163-L173